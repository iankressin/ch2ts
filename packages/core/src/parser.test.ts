import { describe, expect, it } from 'vitest';
import { parse } from './index.js';
import { readFileSync } from 'node:fs';
import { resolve } from 'node:path';

describe('DDL parser (Phase 2)', () => {
  it('parses a simple CREATE TABLE with basic types', () => {
    const sql = readFileSync(resolve(process.cwd(), 'testdata/simple/create_simple.sql'), 'utf8');
    const ast = parse(sql);
    expect(ast).toMatchInlineSnapshot(`
      [
        {
          "columns": [
            {
              "comment": undefined,
              "default": undefined,
              "name": "id",
              "rawType": "UInt64",
              "type": {
                "args": [],
                "name": "UInt64",
              },
            },
            {
              "comment": undefined,
              "default": undefined,
              "name": "name",
              "rawType": "String",
              "type": {
                "args": [],
                "name": "String",
              },
            },
          ],
          "name": "events",
          "orderBy": undefined,
          "partitionBy": undefined,
        },
      ]
    `);
  });

  it('parses nested wrappers and params', () => {
    const sql = `CREATE TABLE db.complicated (
      a Nullable(UInt64) COMMENT 'id',
      b LowCardinality(String),
      c Array(Nullable(Decimal(38,10))),
      d Enum8('A' = 1, 'B' = 2),
      e Tuple(String, UInt32),
      f Map(String, UInt8)
    )`;
    const ast = parse(sql);
    expect(ast[0]?.columns.length).toBe(6);
    expect(JSON.parse(JSON.stringify(ast))).toMatchSnapshot();
  });
});
