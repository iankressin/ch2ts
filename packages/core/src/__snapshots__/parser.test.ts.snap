// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DDL parser (Phase 2) > parses nested wrappers and params 1`] = `
[
  {
    "columns": [
      {
        "comment": "id",
        "name": "a",
        "rawType": "Nullable(UInt64)",
        "type": {
          "args": [
            {
              "args": [],
              "name": "UInt64",
            },
          ],
          "name": "Nullable",
        },
      },
      {
        "name": "b",
        "rawType": "LowCardinality(String)",
        "type": {
          "args": [
            {
              "args": [],
              "name": "String",
            },
          ],
          "name": "LowCardinality",
        },
      },
      {
        "name": "c",
        "rawType": "Array(Nullable(Decimal(38,10)))",
        "type": {
          "args": [
            {
              "args": [
                {
                  "args": [
                    38,
                    10,
                  ],
                  "name": "Decimal",
                },
              ],
              "name": "Nullable",
            },
          ],
          "name": "Array",
        },
      },
      {
        "name": "d",
        "rawType": "Enum8('A'=1,'B'=2)",
        "type": {
          "args": [
            {
              "key": "A",
              "value": 1,
            },
            {
              "key": "B",
              "value": 2,
            },
          ],
          "name": "Enum8",
        },
      },
      {
        "name": "e",
        "rawType": "Tuple(String,UInt32)",
        "type": {
          "args": [
            {
              "args": [],
              "name": "String",
            },
            {
              "args": [],
              "name": "UInt32",
            },
          ],
          "name": "Tuple",
        },
      },
      {
        "name": "f",
        "rawType": "Map(String,UInt8)",
        "type": {
          "args": [
            {
              "args": [],
              "name": "String",
            },
            {
              "args": [],
              "name": "UInt8",
            },
          ],
          "name": "Map",
        },
      },
    ],
    "name": "complicated",
  },
]
`;
